### ISSUE TITLE
Suggestion :: Write-Verbose Behaviour when packaged in a module  

### ISSUE THREAD

**Issue by JohnLBevan** [PowerShell/PowerShell#3106]
When Write-Verbose is included in a function or cmdlet in a module file, it doesn't behave in the same way as it would had that function been written in the main script file.  i.e.
```powershell
# MyScript.ps1
[CmdletBinding()]
param()
 
function Invoke-Demo {
    [CmdletBinding()]
    param ([Parameter()][string]$Message)
    process {Write-Verbose $Message}
}
 
Invoke-Demo 'This DOES show when I run ".\MyScript.ps1 -Verbose"'
Write-Verbose 'This DOES show when I run ".\MyScript.ps1 -Verbose"'
```
Behaves differently to:
```powershell
# MyModule.psm1
function Invoke-Demo {
    [CmdletBinding()]
    param ([Parameter()][string]$Message)
    process {Write-Verbose $Message}
}
```
```powershell
# MyScript.ps1
[CmdletBinding()]
param()
Import-Module -Path '.\MyModule.psm1' -Force
Invoke-Demo 'This does NOT show when I run ".\MyScript.ps1 -Verbose"'
Write-Verbose 'This DOES show when I run ".\MyScript.ps1 -Verbose"'
```
This seems counter intuitive, and overly complex to get write-verbose to function based on whether the script is being run with the verbose switch enabled.

---
**Suggested Solution**
Since there may be a design reason for this behaviour, it's likely best that any new behaviour be controlled explicitly.  As such I'd suggest enabling `Write-Verbose` to function as desired by including a switch when importing the module; i.e. 
```powershell
Import-Module -Path '.\MyModule.psm1' -Force -InheritVerbose
```

---

More information here: https://developer42.wordpress.com/2017/02/04/powershell-suggestion-simplify-write-verbose-in-modules/

**Comment by joeypiccola** [comment]
I hit this a few years back, was it was figured out.

**Comment by mklement0** [comment]
It took a while to emerge, but this is actually the more fundamental issue discussed in #4568:

Inheritance of common parameters - not just `-Verbose`, but notably also `-WhatIf` - via advanced functions relies on _implicitly set local preference variables_: 

A _common parameter_, such as `-Verbose`, passed to an advanced function is function-internally automatically translated into an equivalent local preference variable_,  such as `$VerbosePreference`.

This inheritance mechanism ONLY works:
 * when invoking _binary cmdlets_, which do see and respect the local preference variable.
 * when calling scripts / other functions _in the same module scope as the caller_.

By contrast, a function from a _different_ module _cannot_ see the caller's local preference variable, due to how module scopes function fundamentally.

In short: the common-parameter inheritance mechanism is fundamentally broken for advanced functions across module scopes.

**Comment by ThomasNieto** [comment]
@StevenBucher98 should this be closed as a duplicate/related to #4568?
